flink-connectors/flink-connector-rabbitmq2/src/test/java/org/apache/flink/connector/rabbitmq2/source/RabbitMQSourceTest.java

```diff
@@ -0,0 +1,184 @@
+        addCollectorSink(stream, latch);
+        env.executeAsync();
+
+        // use messages as correlation ids here
+        sendToRabbit(messages, messages);
+
+        latch.await();
+
+        List<String> collectedMessages = getCollectedSinkMessages();
+        assertEquals(messages, collectedMessages);
+    }
+
+    @Test
+    public void exactlyOnceFilterCorrelationIdsTest() throws Exception {
+        List<String> messages = getRandomMessages(5);
+        CountDownLatch latch = new CountDownLatch(3);
+
+        env.enableCheckpointing(5000);
+        DataStream<String> stream = addSourceOn(env, ConsistencyMode.EXACTLY_ONCE);
+        addCollectorSink(stream, latch);
+        env.executeAsync();
+
+        List<String> correlationIds = Arrays.asList("1", "2", "3", "3", "3");
+        sendToRabbit(messages, correlationIds);
+
+        latch.await();
+
+        List<String> collectedMessages = getCollectedSinkMessages();
+        List<String> expectedMessages = messages.subList(0, 3);
+        assertEquals(expectedMessages, collectedMessages);
+    }
+
+    @Test
+    public void exactlyOnceWithFailureAndMessageDuplicationTest() throws Exception {
+        // An exception is thrown in order to trigger a restart of Flink and it
+        // is assured that the system receives the messages only once. We disable
+        // (by setting the interval higher than the test duration) checkpoint to
+        // expect receiving all pre-exception messages once again.
+        env.enableCheckpointing(500000);
+        DataStream<String> stream = addSourceOn(env, ConsistencyMode.EXACTLY_ONCE);
+
+        List<String> messages = getRandomMessages(100);
+        List<String> correlationIds = messages;
+
+        int failAtNthMessage = 30;
+        CountDownLatch latch = new CountDownLatch(messages.size() + failAtNthMessage - 1);
+        addCollectorSink(stream, latch, failAtNthMessage);
+        env.executeAsync();
+
+        sendToRabbit(messages, correlationIds);
+        latch.await();
+
+        List<String> collectedMessages = getCollectedSinkMessages();
+        collectedMessages =
+                collectedMessages.subList(failAtNthMessage - 1, collectedMessages.size());
+        assertEquals(messages, collectedMessages);
+    }
+
+    @Test
+    public void exactlyOnceWithFailureTest() throws Exception {
+        env.enableCheckpointing(1000);
+        DataStream<String> stream = addSourceOn(env, ConsistencyMode.EXACTLY_ONCE);
+
+        List<String> messages = getSequentialMessages(60);
+        List<String> messagesA = messages.subList(0, 30);
+        List<String> messagesB = messages.subList(30, messages.size());
+
+        int failAtNthMessage = messagesA.size() + 1;
+        CountDownLatch latch = new CountDownLatch(messagesA.size() + messagesB.size());
+        addCollectorSink(stream, latch, failAtNthMessage);
+        env.executeAsync();
+
+        sendToRabbit(messagesA, messagesA);
+        TimeUnit.MILLISECONDS.sleep(2500);
```

Comment:
Unfortunately you have to rework this timeout because they are unreliable. I am not sure overall what this test does in addition to `exactlyOnceWithFailureAndMessageDuplicationTest`.